ctms/acoustic_service.py

```diff
@@ -0,0 +1,279 @@
+            if contact_attr_name in acceptable_subdicts and subdict_value is not None:
+                for inner_attr in subdict_value.dict():
+                    inner_value = getattr(subdict_value, inner_attr)
+                    inner_attr_name = str(inner_attr)
+                    if contact_attr_name != "email":
+                        acoustic_field_name = contact_attr_name + "_" + inner_attr_name
+                    else:
+                        # Email table records should not be prefixed
+                        acoustic_field_name = inner_attr_name
+
+                    special_key = (contact_attr_name, inner_attr_name)
+                    if special_key in special_cases:
+                        acoustic_field_name = special_cases.get(special_key)
+
+                    if (
+                        acoustic_field_name
+                        in AcousticResources.VALID_ACOUSTIC_MAIN_TABLE_FIELDS
+                    ):
+                        acoustic_main_table[
+                            acoustic_field_name
+                        ] = self.transform_field_for_acoustic(inner_value)
+                    # else:
+                    # print(f"Skipping CTMS field ({contact_attr}, {inner_attr}) because no match "
+                    #       f"in Acoustic") # TODO: Convert to logging
+        return acoustic_main_table
+
+    def _newsletter_converter(self, acoustic_main_table, contact):
+        # create the RT rows for the newsletter table in acoustic
+        newsletter_rows = []
+        contact_newsletters: List[NewsletterSchema] = contact.newsletters
+        for newsletter in contact_newsletters:
+            if newsletter.name in AcousticResources.MAIN_TABLE_SUBSCR_FLAGS.keys():
+                newsletter_dict = newsletter.dict()
+                create_timestamp = datetime.date.today().isoformat()
+                if "create_timestamp" in newsletter_dict:
+                    create_timestamp = newsletter_dict["create_timestamp"]
+                update_timestamp = datetime.date.today().isoformat()
+                if "update_timestamp" in newsletter_dict:
+                    update_timestamp = newsletter_dict["update_timestamp"]
+                newsletter_template = {
+                    "email_id": contact.email.email_id,
+                    "newsletter_name": newsletter.name,
+                    "newsletter_format": newsletter.format,
+                    "newsletter_lang": newsletter.lang,
+                    "newsletter_source": newsletter.source,
+                    "newsletter_unsub_reason": newsletter.unsub_reason,
+                    "create_date": create_timestamp,
+                    "update_date": update_timestamp,
+                }
+                newsletter_rows.append(newsletter_template)
+                # and finally flip the main table's sub_<newsletter> flag to true for each subscription
+                if newsletter.subscribed:
+                    acoustic_main_table[
```

Comment:
is "acoustic_main_table" an accessible dict in this def?

Reply:
I guess it is, huh. But it doesn't get passed back out. Is that weird?

Reply:
It was provided as a mutable dictionary... but yeah, that could be a bit weird. I can modify the behavior to return both dictionaries